contracts/src/v0.8/PermissionedForwardProxy.sol

```diff
@@ -0,0 +1,58 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/utils/Address.sol";
+import "./ConfirmedOwner.sol";
+
+/**
+ * @title PermissionedForwardProxy
+ * @notice This proxy is used to forward calls from sender to target. It maintains
+ * a permission list to check which sender is allowed to call which target
+ */
+contract PermissionedForwardProxy is ConfirmedOwner {
+  using Address for address;
+
+  error PermissionNotSet();
+
+  mapping(address => address) private s_forwardPermissionList;
+
+  constructor() ConfirmedOwner(msg.sender) {}
+
+  /**
+   * @notice Verifies if msg.sender has permission to forward to target address and then forwards the handler
+   * @param target address of the contract to forward the handler to
+   * @param handler bytes to be passed to target in call data
+   */
+  function forward(address target, bytes calldata handler) external {
+    if (s_forwardPermissionList[msg.sender] != target) {
+      revert PermissionNotSet();
+    }
+    target.functionCall(handler);
+  }
+
+  /**
+   * @notice Adds permission for sender to forward calls to target via this proxy.
+   * Note that it allows to overwrite an existing permission
+   * @param sender The address who will use this proxy to forward calls
+   * @param target The address where sender will be allowed to forward calls
+   */
+  function setPermission(address sender, address target) external onlyOwner {
+    s_forwardPermissionList[sender] = target;
```

Comment:
Does it make sense to emit an event here (same for functions below)?

Reply:
Yeah I think those will be helpful. Will add them